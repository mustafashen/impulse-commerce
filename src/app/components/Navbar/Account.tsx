
export default function Account() {
  
  function logInCheck(): boolean {
    return true
  }
  // TODO: Make these two lists individual components
  const loggedIn = <ul className="invisible group-hover:visible group-focus-within:visible">
    <li>__name__</li>
    <li>Orders</li>
    <li>Favorites</li>
    <li>Account Settings</li>
    <li>Logout</li>
  </ul>

  const notLoggedIn = <ul className="invisible group-hover:visible group-focus-within:visible">
    <li>Login</li>
    <li>Sign Up</li>
  </ul>

  return (
    <li className="group w-24">
        <button className="h-[--navbar-h] flex flex-col justify-center w-full">Account</button>
        {logInCheck() ? loggedIn : notLoggedIn}
    </li>
  )
}
