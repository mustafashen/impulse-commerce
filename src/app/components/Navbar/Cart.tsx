import { useNavbarUserContext } from "./context/NavbarUserContext"

export default function Cart() {
  const {cartItems} = useNavbarUserContext()
  return (
    <li className="group w-24">
        <button className="h-[--navbar-h] flex flex-col justify-center w-full">Cart</button>
        <ul className="invisible group-hover:visible group-focus-within:visible">
          {
            cartItems.map((item: {productID: string, productName: string}) => {
              return <li key={item.productID}>
                <span>{item.productID}</span>
                <span>{item.productName}</span>
              </li>
            })
          }
        </ul>
    </li>
  )
}
