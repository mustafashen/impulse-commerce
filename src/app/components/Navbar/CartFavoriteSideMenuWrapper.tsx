import CartSideMenu from "./CartSideMenu";
import FavoritesSideMenu from "./FavoritesSideMenu";
import { useCartFavViewContext } from "@/app/contexts/CartFavViewContext";

export default function CartFavoriteSideMenuWrapper() {
  const {cartFavView, setCartFavView} = useCartFavViewContext()

  function menuSelection(view: string) {
    switch (view) {
      case 'closed':
        return false
      case 'cart':
        return (    
            <CartSideMenu/>
        )
      case 'fav':
        return (    
          <FavoritesSideMenu/>
        )
      default:
        break;
    }
  }

  function handleClose() {
    setCartFavView('closed')
  }

  return (
    <div className={`fixed transition-all top-[--navbar-h] z-50 ${cartFavView === 'closed' ? '-right-[100%]' : 'right-0'}`}>
      <div>
        <button onClick={handleClose}>Close</button>
      </div>
      {menuSelection(cartFavView)}
    </div>
  )
}
