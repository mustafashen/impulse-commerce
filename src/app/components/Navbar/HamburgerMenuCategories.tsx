import Link from "next/link";
import { useCategoriesContext } from "@/app/components/Navbar/context/CategoriesContext";
import { Suspense } from "react";
import Skeleton from "react-loading-skeleton";

type CategoryType = {
  id: string,
  name: string,
  subCategories?: CategoryType[]
}

export default function HamburgerMenuCategories(): React.ReactElement {
  const {categories} = useCategoriesContext()

  function ParseCategories({category, layer = 0}: {category: CategoryType, layer?: number}): JSX.Element {
    const { name, id, subCategories } = category

    // Give each parent container group name,
    // and each of their children list visible on focus attribute
    // This has to defined as a distinct value because tailwind compiler cannot handle indistinct values
    let [layerGroup, layerInvisibleElement]: [string, string] = ["", ""] 
    switch (layer) {
      case 0:
        [layerGroup, layerInvisibleElement] = [`group/cat0`,`group-focus-within/cat0:visible`]
        break;
      case 1:
        [layerGroup, layerInvisibleElement] = [`group/cat1`,`group-focus-within/cat1:visible`]
        break;
      case 2:
        [layerGroup, layerInvisibleElement] = [`group/cat2`,`group-focus-within/cat2:visible`]
        break;
      case 3:
        [layerGroup, layerInvisibleElement] = [`group/cat3`,`group-focus-within/cat3:visible`]
        break;
      default:
        break;
    }

    const ListElementSubCategoryStyle: string = `h-[--sub-cat-h]`
    const ListElementMainCategoryStyle: string = `h-16`
    const ListElementCategoryStyle: string = `flex flex-row flex-nowrap overflow-visible ${layerGroup}`
    const ListElementStyle: string = `${layer === 0 ? ListElementMainCategoryStyle : ListElementSubCategoryStyle} ${ListElementCategoryStyle}`

    const ChildListMainCategoryStyle: string = `whitespace-nowrap top-[--vertical-sub-cat-pos]`
    const ChildListSubCategoryStyle: string = `top-0`
    const ChildListStyle: () => string = () => `${layer === 0 ? ChildListMainCategoryStyle : ChildListSubCategoryStyle} absolute left-36 invisible ${layerInvisibleElement}`

    // If there are children for the category
    // Create a category list element,
    // inside it, create an unorganized list,
    // recall the function inside the list
    if (subCategories) {
      return <li className={ListElementStyle} key={id}>
        <button className="h-[--sub-cat-h]">
          {`${name}>`}
        </button>
        <ul className={ChildListStyle()}>
            {subCategories.map((el: CategoryType) => {
              return ParseCategories({category: el, layer: ++layer})
            })}
            <li className="h-[--sub-cat-h] flex flex-col justify-center">
                <Link
                  href={{
                  pathname: `/categories/${id}`,
                  query: {name}
                  }}>
                  {`[All ${name}]`}
                </Link>
            </li>
        </ul>
      </li>
    }

    // If children has no sub category, return a category list element
    return <li key={id} className="h-[--sub-cat-h] flex flex-col justify-center">
        <Link
          href={{
          pathname: `/categories/${id}`,
          query: {name}
          }}>
          {name}
        </Link>
    </li>
  }

  return (
    <>
      {
        categories.subCategories ?
          <Suspense fallback={<Skeleton/>}>
            {categories.subCategories.map((el: CategoryType) => (
              <ParseCategories key={el.id} category={el} />
            ))}
          </Suspense>
        :
        <div>No categories available.</div>
      }
    </>
  )
}

