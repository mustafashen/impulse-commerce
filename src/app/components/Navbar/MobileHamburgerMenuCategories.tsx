"use client"

import Link from "next/link";
import { useCategoriesContext } from "@/app/components/Navbar/context/CategoriesContext";
import { Suspense, useEffect, useState } from "react";
import Skeleton from "react-loading-skeleton";

type CategoryType = {
  id: string,
  name: string,
  subCategories?: CategoryType[]
}

export default function MobileHamburgerMenuCategories(): React.ReactElement {
  const {categories} = useCategoriesContext()
  const [groupVisible, setGroupVisible] = useState<{[id: string]: string}>({})
  useEffect(() => setGroupVisible(generateGroupTree(categories)),[])
  
  function generateGroupTree(categories: CategoryType) {
		// start from the root
    // check if category has children
    // if it has, give it a group visibility state
    // iterate each of its children
    // recall the function on them

		let groupVisibility: {[id: string]: string} = {}
		function traverseTree(category: CategoryType) {
			const { id, subCategories } = category

			if (subCategories) {
        groupVisibility[id] = "hidden"
        subCategories.forEach((subCategory: CategoryType) => {
          traverseTree(subCategory)
        })
      }
    }
    
    traverseTree(categories)
    groupVisibility.all = "table-row"

		return groupVisibility
	}

  function switchGroupView(groupID: string) {
    let newState = {...groupVisible}
    
    for (const key of Object.keys(newState)) {
        if (newState[key]) {
          if (key == groupID) {
            newState[key] = "table-row"
          } else {
            newState[key] = "hidden"
          }
        }
    }
    setGroupVisible(newState)
  }

  function handleViewSwitch(groupID: string) {
    switchGroupView(groupID)
  }

	function ParseCategories(category: CategoryType): JSX.Element[] {
    // create a tr array
    // Start from root category, pass the parent category
    // Check if there is "subCategories"
    // create a tr
      // create a td to show current group
      // create a td for going back to previous view
      // iterate subCategories
      // call this function with subCategory and the parent of the subCategory
      // if children has children put a button inside that references that group
      // else put a link in it for its page
    // push the tr to the tr array
    
    const groupRows: JSX.Element[] = []
    function generateCategoryGroups(category: CategoryType, parentGroup: CategoryType): void {
      const {subCategories, id, name} = category

      if (subCategories) {
        groupRows.push(
          <tr key={id} className={`${groupVisible[id]}`}>
            <td className="block"><b>{name}</b></td>
            <td className="block">
              <button onClick={() => {handleViewSwitch(parentGroup.id)}}>{`<${parentGroup.name}`}</button>
            </td>

            {
              subCategories.map((subCategory: CategoryType) => {
                const {id: subID, name: subName, subCategories: subSubCategories} = subCategory
                generateCategoryGroups(subCategory, category)
                if (subSubCategories) {
                  return <td key={subID} className="block">
                    <button onClick={() => {handleViewSwitch(subID)}}>
                      {`${subName}>`}
                    </button>
                  </td>
                } else {
                  return <td key={subID} className="block">
                    <Link
                    href={{
                    pathname: `/categories/${id}`,
                    query: {subName}
                    }}>
                      {subName}
                    </Link>
                  </td>
                }
              })
            }

            <td className="block">
              <Link
              href={{
              pathname: `/categories/${id}`,
              query: {name}
              }}>
                {`[${name}]`}
              </Link>
            </td>
          </tr>
        )
      }
      
    }

    generateCategoryGroups(category, category)
    
    return groupRows
  }

  return (
    <table className = {`z-40 lg:hidden`}>
    <thead></thead>
    <tbody>
      {
        groupVisible ? 
          <Suspense fallback={<Skeleton/>}>
            {ParseCategories(categories).map((el: JSX.Element) => {
              return el
            })}
          </Suspense>
          : 
          <div>No categories available.</div>
      }
    </tbody>
    </table>
  )
}
