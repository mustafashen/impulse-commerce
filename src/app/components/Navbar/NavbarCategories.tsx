import Link from "next/link";
import { useCategoriesContext } from "@/app/components/Navbar/context/CategoriesContext";
import { Suspense } from "react";
import Skeleton from "react-loading-skeleton";

type CategoryType = {
  id: string,
  name: string,
  subCategories?: CategoryType[]
}

export default function NavbarCategories(): React.ReactElement {
  const {categories} = useCategoriesContext()
  
  function ParseNavbarCategories(category: CategoryType, layer: number = 0): JSX.Element {
    const { name, id, subCategories } = category

    let [layerGroup, layerInvisibleElement]: [string, string] = ["", ""]
    switch (layer) {
      case 0:
        [layerGroup, layerInvisibleElement] = [`group/cat0`,`group-hover/cat0:visible group-focus-within/cat0:visible`]
        break;
      case 1:
        [layerGroup, layerInvisibleElement] = [`group/cat1`,`group-hover/cat1:visible group-focus-within/cat1:visible`]
        break;
      case 2:
        [layerGroup, layerInvisibleElement] = [`group/cat2`,`group-hover/cat2:visible group-focus-within/cat2:visible`]
        break;
      case 3:
        [layerGroup, layerInvisibleElement] = [`group/cat3`,`group-hover/cat3:visible group-focus-within/cat3:visible`]
        break;
      default:
        break;
    }

    const ListElementSubCategoryStyle: string = `flex flex-row flex-nowrap ${layerGroup} h-[--sub-cat-h]`
    const ListElementMainCategoryStyle: string = `${layerGroup}`
    const ListElementCategoryStyle: string = `w-24`
    const ListElementStyle: string = `${layer === 0 ? ListElementMainCategoryStyle : ListElementSubCategoryStyle} ${ListElementCategoryStyle}`

    const ButtonMainCategoryStyle: string = `first:h-[--navbar-h] flex flex-col justify-center w-full`
    const ButtonStyle: () => string = () => layer === 0 ?  ButtonMainCategoryStyle : "h-[--sub-cat-h]"

    const ChildListMainCategoryStyle: string = `invisible group ${layerInvisibleElement} whitespace-nowrap`
    const ChildListSubCategoryStyle: string = `invisible ${layerInvisibleElement}`
    const ChildListStyle: () => string = () => layer === 0 ? ChildListMainCategoryStyle : ChildListSubCategoryStyle

    if (subCategories) {
      return <li className={ListElementStyle} key={id}>
          <button className={ButtonStyle()}>
            {`${name}>`}
          </button>
          <ul className={ChildListStyle()}>
            {subCategories.map((el: CategoryType) => {
              return ParseNavbarCategories(el, ++layer)
            })}
            <li className="h-[--sub-cat-h] flex flex-col justify-center">
              <Link
                href={{
                pathname: `/categories/${id}`,
                query: {name}
                }}>
                {`[All ${name}]`}
              </Link>
            </li>
          </ul>
        
        </li>
    }

    return <li key={id} className="h-[--sub-cat-h] flex flex-col justify-center">
      <div>
        <Link
          href={{
          pathname: `/categories/${id}`,
          query: {name}
          }}>
          {name}
        </Link>
      </div>
    </li>
  }

  return (
    <>
      <ul className={"list-none flex flex-row flex-nowrap space-x-5 justify-center h-full"}>
        {
          categories.subCategories ?
            <Suspense fallback={<Skeleton/>}>
              {
                categories.subCategories.map((category: CategoryType): React.ReactElement => {
                  return ParseNavbarCategories(category)
                })
              }
            </Suspense>  
          :
            <div>No categories available.</div>
        }
      </ul>
    </>
  )
}
