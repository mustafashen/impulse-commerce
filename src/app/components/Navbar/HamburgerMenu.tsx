"use client"

import Image from "next/image"
import { useState } from "react"
import MobileHamburgerMenuCategories from "@/app/components/Navbar/MobileHamburgerMenuCategories"
import HamburgerMenuCategories from "@/app/components/Navbar/HamburgerMenuCategories"


export default function HamburgerMenu(): React.ReactElement {
  const [menuVisible, setMenuVisible] = useState<{isVisible: string}>({isVisible: "invisible"})


  const handleMouseEnter = () => setMenuVisible({isVisible: "visible"})
  
  function handleMouseLeave() {
    setMenuVisible({isVisible: "invisible [&>*]:outline-none"})

    if (document.activeElement instanceof HTMLElement) {
      document.activeElement.blur()
    }
  }

  function handleClick() {
    if (menuVisible.isVisible === "invisible") setMenuVisible({isVisible: "visible"})
    else {
      setMenuVisible({isVisible: "invisible"})
      
      if (document.activeElement instanceof HTMLElement) {
        document.activeElement.blur()
      }
    }
  }


  return (
      <li id="hamburger-menu" className="h-48 w-48 " onMouseEnter={handleMouseEnter} onMouseLeave={handleMouseLeave}>
        <div id="hm-icon-wrap" className={`flex flex-col justify-center items-center h-[--navbar-h] w-12`} onClick={handleClick}>
              <Image  src="/icons/hamburger.svg"
                      width={30}
                      height={1}
                      sizes=""
                      alt="hamburger menu icon"
                      >
              </Image>
        </div>
        <ul id="hm-ctg-wrap" className = {`h-[--vertical-menu-h] w-[--vertical-menu-w] pt-[--vertical-menu-pt] z-40 max-lg:hidden ${menuVisible.isVisible}`}>
          <HamburgerMenuCategories/>
        </ul>
        <div id="mhm-ctg-wrap" className={`${menuVisible.isVisible}`}>
          <MobileHamburgerMenuCategories/>
        </div>
      </li>
  )
}
