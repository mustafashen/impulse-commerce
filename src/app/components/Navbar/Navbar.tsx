import Logo from "@/app/components/Navbar/Logo"
import HamburgerMenu from "@/app/components/Navbar/HamburgerMenu"
import NavbarUserSection from "@/app/components/Navbar/NavbarUserSection"
import NavbarCategories from "@/app/components/Navbar/NavbarCategories"
import { CategoriesContextProvider } from "./context/CategoriesContext"


export default function Navbar({logoInMiddle = false, verticalMenu= false, categoryInMiddle= false}): React.ReactElement {
  
  const VerticalMenuLayout: JSX.Element = <>
      <HamburgerMenu/>
      <li id="logo-middle" className={`absolute w-full top-[--logo-top] left-0 flex justify-center h-0`}><Logo/></li>
      <li id="navbar-user-section"><NavbarUserSection/></li>
  </>

  const LogoInMiddleLayout: JSX.Element = <>
    <li id="navbar-categories" className="navbar-categories"><NavbarCategories/></li>
    <li id="logo-middle" className={`absolute w-full top-[--logo-top] left-0 flex justify-center h-0`}><Logo/></li>
    <li id="navbar-user-section"><NavbarUserSection/></li>
  </>

  const LogoInLeftLayout: JSX.Element = <>
    <div className="flex flex-row flex-nowrap">
        <li id="logo-left" className={`flex justify-center items-center`}><Logo/></li>
        <li id="navbar-categories" className="navbar-categories"><NavbarCategories/></li>
    </div>
    <li id="navbar-user-section"><NavbarUserSection/></li>
  </>

  const MobileLayout: JSX.Element = <>
    <HamburgerMenu/>
    <li id="logo-middle" className={`absolute w-full top-[--logo-top] left-0 flex justify-center h-0`}><Logo/></li>
  </>

  const CategoryMiddleLayout: JSX.Element = <>
    <li id="logo-left" className={`flex justify-center items-center`}><Logo/></li>
    <li id="navbar-categories-middle" className={`absolute w-full left-0 flex justify-center h-0`}><NavbarCategories/></li>
    <li id="navbar-user-section"><NavbarUserSection/></li>
  </>

  function LayoutSelection(): JSX.Element {
    if (verticalMenu) {
      return VerticalMenuLayout
    } else if (logoInMiddle) {
      return LogoInMiddleLayout
    } else if(categoryInMiddle) {
      return CategoryMiddleLayout
    } else return LogoInLeftLayout
  }

  return (
    <CategoriesContextProvider>
      <nav className={`h-[--navbar-h] w-screen border-solid border-4 border-stone-900 top-0 overflow-visible box-border z-40 fixed`}>
          <ul className={"flex flex-row flex-nowrap w-full h-full justify-between max-lg:hidden"}>
                {LayoutSelection()}
          </ul>
          <ul className={"flex-row flex-nowrap w-full h-full justify-between hidden max-lg:flex"}>
              {MobileLayout}
          </ul>
      </nav>   
    </CategoriesContextProvider>
  )
}
