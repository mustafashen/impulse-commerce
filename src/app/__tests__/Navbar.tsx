import { fireEvent, getByRole, render, screen } from '@testing-library/react'
import Navbar from '@/app/components/Navbar/Navbar'
import '@testing-library/jest-dom'
import HamburgerMenu from '@/app/components/Navbar/HamburgerMenu'
import HamburgerMenuCategories from '@/app/components/Navbar/HamburgerMenuCategories'
import {CategoriesContextProvider} from '@/app/test-utils/TestCategoriesContext'
import {getMockCategoryNames} from '@/app/test-utils/getMockCategory'
import NavbarCategories from '@/app/components/Navbar/NavbarCategories'
import MobileHamburgerMenuCategories from '../components/Navbar/MobileHamburgerMenuCategories'

describe('Navbar', () => {
  
	it ('renders a navbar', () => {
    render(<Navbar />)

    const navbar = screen.getByRole('navigation')

    expect(navbar).toBeInTheDocument()
  })

  it ('renders navbar in logo in middle layout', () => {
    const {container} = render(<Navbar logoInMiddle={true}/>)

    const navbarUserSection = container.querySelector('#navbar-user-section')
    expect(navbarUserSection).toBeInTheDocument()

    const navbarCategories = container.querySelector('#navbar-categories')
    expect(navbarCategories).toBeInTheDocument()

    const logoInMiddle = container.querySelector('#logo-middle')
    expect(logoInMiddle).toBeInTheDocument()

  })

  it ('renders navbar in logo in left layout', () => {
    const {container} = render(<Navbar logoInMiddle={false}/>)

    const navbarUserSection = container.querySelector('#navbar-user-section')
    expect(navbarUserSection).toBeInTheDocument()

    const navbarCategories = container.querySelector('#navbar-categories')
    expect(navbarCategories).toBeInTheDocument()

    const logoInLeft = container.querySelector('#logo-left')
    expect(logoInLeft).toBeInTheDocument()
  })

  it ('renders navbar in vertical menu layout', () => {
    const {container} = render(<Navbar verticalMenu={true}/>)
    
    const navbarUserSection = container.querySelector('#navbar-user-section')
    expect(navbarUserSection).toBeInTheDocument()

    const hamburgerMenu = container.querySelector('#hamburger-menu')
    expect(hamburgerMenu).toBeInTheDocument()

    const logoInMiddle = container.querySelector('#logo-middle')
    expect(logoInMiddle).toBeInTheDocument()

  })
 
  it ('renders navbar in category middle layout', () => {
    const {container} = render(<Navbar categoryInMiddle={true}/>)

    const navbarUserSection = container.querySelector('#navbar-user-section')
    expect(navbarUserSection).toBeInTheDocument()

    const navbarCategoriesMiddle = container.querySelector('#navbar-categories-middle')
    expect(navbarCategoriesMiddle).toBeInTheDocument()

    const logoInLeft = container.querySelector('#logo-left')
    expect(logoInLeft).toBeInTheDocument()
  })

})

describe('NavbarCategories', () => {
  it ('NavbarCategories with context renders categories', () => {
    render(<NavbarCategories/>, {wrapper: CategoriesContextProvider})
    
    getMockCategoryNames().forEach(async (ctg) => {
      const item = await screen.findByRole('listitem', {name: new RegExp(ctg.name)})
      expect(item).toBeInTheDocument()
    })
    
  }) 
})

describe ('HamburgerMenu', () => {
  it ('makes HamburgerMenuCategories wrapper visible/invisible on mouseEnter/mouseLeave', () => {
    const {container} = render(<HamburgerMenu/>)

    const hamburgerMenuCategories = container.querySelector('#hm-ctg-wrap')
    const hamburgerMenu = container.querySelector('#hamburger-menu')

    expect(hamburgerMenuCategories).toHaveClass('invisible')

    if (hamburgerMenu) {
      fireEvent.mouseOver(hamburgerMenu)
    }
    
    expect(hamburgerMenuCategories).toHaveClass('visible')

  }) 

  it ('makes HamburgerMenuCategories wrapper visible/invisible on mouseClick', () => {
    const {container} = render(<HamburgerMenu/>)

    const hamburgerMenuCategories = container.querySelector('#hm-ctg-wrap')
    const hamburgerMenu = container.querySelector('#hm-icon-wrap')

    expect(hamburgerMenuCategories).toHaveClass('invisible')

    if (hamburgerMenu) {
      fireEvent.click(hamburgerMenu)
    }
    expect(hamburgerMenuCategories).toHaveClass('visible')

    if (hamburgerMenu) {
      fireEvent.click(hamburgerMenu)
    }
    expect(hamburgerMenuCategories).toHaveClass('invisible')
  })

  it ('makes MobileHamburgerMenuCategories wrapper visible/invisible on tap', () => {
    const {container} = render(<HamburgerMenu/>)

    const hamburgerMobileMenuCategories = container.querySelector('#mhm-ctg-wrap')

    const hamburgerMenu = container.querySelector('#hm-icon-wrap')

    expect(hamburgerMobileMenuCategories).toHaveClass('invisible')

    if (hamburgerMenu) {
      fireEvent.click(hamburgerMenu)
    }
    expect(hamburgerMobileMenuCategories).toHaveClass('visible')

    if (hamburgerMenu) {
      fireEvent.click(hamburgerMenu)
    }
    expect(hamburgerMobileMenuCategories).toHaveClass('invisible')
  })
  
})



describe('HamburgerMenuCategories', () => {

  it ('HamburgerMenuCategories with no context renders a message div', () => {
    const {container} = render(<HamburgerMenuCategories/>)

    expect(container).toMatchSnapshot(`
      <>
        <div>No categories available.</div>
      </>
    `)
  }) 

  it ('HamburgerMenuCategories with context renders categories', () => {
    render(<HamburgerMenuCategories/>, {wrapper: CategoriesContextProvider})
    
    getMockCategoryNames().forEach(async (ctg) => {
      const item = await screen.findByRole('listitem', {name: new RegExp(ctg.name)})
      expect(item).toBeInTheDocument()
    })
    
  }) 

})

describe('HamburgerMenuCategories', () => {

  it ('HamburgerMobileMenuCategories with no context renders a message div', () => {
    const {container} = render(<MobileHamburgerMenuCategories/>)

    expect(container).toMatchSnapshot(`
      <>
        <div>No categories available.</div>
      </>
    `)
  }) 

  it ('HamburgerMobileMenuCategories with context renders categories', () => {
    render(<MobileHamburgerMenuCategories/>, {wrapper: CategoriesContextProvider})
    
    getMockCategoryNames().forEach(async (ctg) => {
      const item = await screen.findByText(new RegExp(ctg.name))
      expect(item).toBeInTheDocument()
    })
    
  }) 

})