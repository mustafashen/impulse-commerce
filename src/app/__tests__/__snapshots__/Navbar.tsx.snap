// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HamburgerMenuCategories HamburgerMenuCategories with no context renders a message div: 
      <>
        <div>No categories available.</div>
      </>
     1`] = `
<div>
  <div>
    No categories available.
  </div>
</div>
`;

exports[`HamburgerMenuCategories HamburgerMobileMenuCategories with no context renders a message div: 
      <>
        <div>No categories available.</div>
      </>
     1`] = `
<div>
  <table
    class="z-40 lg:hidden"
  >
    <thead />
    <tbody />
  </table>
</div>
`;
